src/terminal/terminal.c

```diff
@@ -1529,7 +1537,56 @@ static int __guac_terminal_send_key(guac_terminal* term, int keysym, int pressed
         /* Typeable keys of number pad */
         else if (keysym >= 0xFFAA && keysym <= 0xFFB9) {
             char value = keysym - 0xFF80;
-            guac_terminal_send_data(term, &value, sizeof(value));
+            if (term->func_keys_and_keypad == GUAC_TERMINAL_FUNC_KEYS_AND_KEYPAD_VT100PLUS &&
+                    term->application_keypad_keys) {
+                /* https://vt100.net/docs/vt100-ug/chapter3.html */
+                switch (value) {
+                    case '0':
+                        return guac_terminal_send_string(term, "\x1BOp");
+                    case '1':
+                        return guac_terminal_send_string(term, "\x1BOq");
+                    case '2':
+                        return guac_terminal_send_string(term, "\x1BOr");
+                    case '3':
+                        return guac_terminal_send_string(term, "\x1BOs");
+                    case '4':
+                        return guac_terminal_send_string(term, "\x1BOt");
+                    case '5':
+                        return guac_terminal_send_string(term, "\x1BOu");
+                    case '6':
+                        return guac_terminal_send_string(term, "\x1BOv");
+                    case '7':
+                        return guac_terminal_send_string(term, "\x1BOw");
+                    case '8':
+                        return guac_terminal_send_string(term, "\x1BOx");
+                    case '9':
+                        return guac_terminal_send_string(term, "\x1BOy");
```

Comment:
Perhaps this could be calculated instead of explicitly listing each mapping from 0-9 onto p-y?